Sources/BraintreePayPal/BTPayPalClient.swift

```diff
@@ -0,0 +1,429 @@
+        }
+    }
+    
+    // MARK: - Private Methods
+    
+    private func performSwitchRequest(
+        appSwitchURL: URL,
+        paymentType: BTPayPalPaymentType,
+        completion: @escaping (BTPayPalAccountNonce?, Error?) -> Void
+    ) {
+        approvalURL = appSwitchURL
+        authenticationSession = ASWebAuthenticationSession(
+            url: appSwitchURL,
+            callbackURLScheme: BTPayPalRequest.callbackURLScheme
+        ) { callbackURL, error in
+                self.authenticationSession = nil
+                if let error = error as? NSError {
+                    if error.domain == ASWebAuthenticationSessionError.errorDomain,
+                       error.code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
+                        if self.returnedToAppAfterPermissionAlert == true {
+                            // User tapped system cancel button in browser
+                            let eventName = "ios.\(paymentType.stringValue).authsession.browser.cancel"
+                            self.apiClient.sendAnalyticsEvent(eventName)
+                        } else {
+                            // User tapped system cancel button on permission alert
+                            let eventName = "ios.\(paymentType.stringValue).authsession.alert.cancel"
+                            self.apiClient.sendAnalyticsEvent(eventName)
+                        }
+                    }
+                    
+                    // User canceled by breaking out of the PayPal browser switch flow
+                    // (e.g. System "Cancel" button on permission alert or browser during ASWebAuthenticationSession)
+                    completion(nil, BTPayPalError.canceled)
+                    return
+                }
+
+                self.handleBrowserSwitchReturn(callbackURL, paymentType: paymentType, completion: completion)
+            }
+        
+        authenticationSession?.presentationContextProvider = self
+        returnedToAppAfterPermissionAlert = false
+        isAuthenticationSessionStarted = authenticationSession?.start() ?? false
+
+        let authenticationSessionStatus = isAuthenticationSessionStarted ? "succeeded" : "failed"
+        apiClient.sendAnalyticsEvent("ios.\(paymentType.stringValue).authsession.start.\(authenticationSessionStatus)")
+    }
+    
+    private func decorate(approvalURL: URL, for request: BTPayPalRequest) -> URL {
+        guard let request = payPalRequest as? BTPayPalCheckoutRequest,
+              var approvalURLComponents = URLComponents(url: approvalURL, resolvingAgainstBaseURL: false) else {
+            return approvalURL
+        }
+
+        let userActionValue = request.userAction.stringValue
+        guard userActionValue.count > 0 else {
+            return approvalURL
+        }
+        
+        let userActionQueryItem = URLQueryItem(name: "useraction", value: userActionValue)
+        var queryItems = approvalURLComponents.queryItems ?? []
```

Comment:
We want to set `var queryItems = approvalURLComponents.queryItems ?? []` to an empty array if it doesn't exist. If we append directly to `approvalURLComponents.queryItems` the `queryItems` are optional and it would not be appended to if it does not exist yet.

Reply:
Yeah it's a little tough to explain -- our tests fail if we write `approvalURLComponents.queryItems?.append(userActionQueryItem)` because in some cases `approvalURLComponents.queryItems` is nil, so when it comes time to make the URL from the components, we have not appended anything to `queryItems`. In the current implementation we set it to an empty array to avoid that problem.